service/core/kafka/server/src/main/scala/com/lightbend/lagom/internal/broker/kafka/TopicProducerActor.scala

```diff
@@ -52,79 +88,108 @@ private[lagom] class TaggedOffsetProducerActor[Message](
+        RestartSource.withBackoff(
+          producerConfig.minBackoff,
+          producerConfig.maxBackoff,
+          producerConfig.randomBackoffFactor
+        ) { () =>
+          val brokersAndOffset: Future[(String, OffsetDao)] = eventualBrokersAndOffset(tagName)
+          Source
+            .fromFuture(brokersAndOffset)
+            .initialTimeout(producerConfig.offsetTimeout)
+            .flatMapConcat {
+              case (endpoints, offset) =>
+                val serviceName = kafkaConfig.serviceName.map(name => s"[$name]").getOrElse("")
+                log.debug("Kafka service {} located at URIs [{}] for producer of [{}]", serviceName, endpoints, topicId)
+                val eventStreamSource: Source[(Message, Offset), _]          = eventStreamFactory(tagName, offset.loadedOffset)
+                val userlandFlow: Flow[(Message, Offset), Future[Done], Any] = eventsPublisherFlow(endpoints, offset)
+                eventStreamSource.via(userlandFlow)
+            }
+        }
+      }
 
-    case (Some(endpoints: String), dao: OffsetDao) =>
-      // log is impacted by kafkaConfig.serviceName
-      val serviceName = kafkaConfig.serviceName.map(name => s"[$name]").getOrElse("")
-      log.debug("Kafka service {} located at URIs [{}] for producer of [{}]", serviceName, endpoints, topicId)
-      run(tag, endpoints, dao)
+      val (killSwitch, streamDone) = backoffSource
+        .viaMat(KillSwitches.single)(Keep.right)
+        .toMat(Sink.ignore)(Keep.both)
+        .run()
 
-    case (None, _) =>
-      // log is impacted by kafkaConfig.serviceName
-      kafkaConfig.serviceName match {
-        case Some(serviceName) => log.error("Unable to locate Kafka service named [{}]", serviceName)
-        case None              => log.error("Unable to locate Kafka brokers URIs")
-      }
+      shutdown = Some(killSwitch)
+      streamDone.pipeTo(self)
+    }
 
-      context.stop(self)
-  }
+    case EnsureActive(_) =>
+    // Yes, we are active
 
-  private def active: Receive = generalHandler.orElse {
     case Done =>
+      // This `Done` is materialization of the `Sink.ignore` above.
       log.info("Kafka producer stream for topic {} was completed.", topicId)
       context.stop(self)
+
+    case Status.Failure(e) =>
+      throw e
   }
 
-  private def run(tag: String, endpoints: String, dao: OffsetDao) = {
-    val readSideSource = eventStreamFactory(tag, dao.loadedOffset)
+  /**
+   * Every time we want to re/start the stream we need to locate the brokers and load the latest offset.
+   * The returned future can fail when the offset store is unavailable or times out or when the brokers list
+   * is empty or unconfigured, etc... In either case, the stream can't be built
+   */
+  private def eventualBrokersAndOffset(tagName: String): Future[(String, OffsetDao)] = {
+    // TODO: review the OffsetStore API. `prepare()` does a lot more than we want here. Ideally, prepare would be
```

Comment:
Is prepare an idempotent operation that we can execute multiple times?

Reply:
It's idempotent but more costly then strictly necessary. I've had this API get on my way in the past already. I'll raise an issue.